### PR TITLE
gpconfig: add confirmation to continue for user when a host is down

### PR DESCRIPTION
**Pull request by divyeshddv**
When a host is down, and a user runs gpconfig to update any config: we should warn them and ask for confirmation to continue with the config update. Since they might get out of sync config values.

The warning and ask for confirmation were part of gpconfig until 6X, which got removed during py2/3 migration change.

This change adds that warning/ask for confirmation back in. 

Tracker story: https://www.pivotaltracker.com/story/show/177336745
Pipeline: https://cm.ci.gpdb.pivotal.io/teams/main/pipelines/7x_fix_gpconfig_down_host

### CODE REVIEW

File: gpMgmt/test/behave/mgmt_utils/gpconfig.feature [hunk 1]
```diff
@@ -179,3 +179,39 @@ Feature: gpconfig integration tests
...
+    @concourse_cluster
+    @demo_cluster
+    Scenario: gpconfig with coordinatoronly update does not ask for confirmation when a host is unreachable
+      Given the database is running
+        And the host for the primary on content 0 is made unreachable
+
+      When the user runs "gpconfig -c max_connections -v 10 --coordinatoronly"
+
+      Then gpconfig should print "completed successfully with parameters '-c max_connections -v 10 --coordinatoronly'" to stdout
+        And gpconfig should return a return code of 0
```
Comment (kalensk) [inline gpMgmt/test/behave/mgmt_utils/gpconfig.feature:217]: Consider adding a line feed at the end of files. This can help with certain editors and prevent the annoying github warning icon as shown here.

In IntelliJ:
Preferences > Editor > General > (at the bottom) Ensure line feed at file end on save

File: gpMgmt/bin/gpconfig [hunk 2]
```diff
@@ -199,7 +200,7 @@ class Guc:
 
 
 def confirm_user_wants_to_continue():
-    if not ask_yesno('', "Are you sure you want to ignore unreachable hosts?", 'N'):
+    if not ask_yesno('', "Are you sure you want to ignore unreachable hosts? Updating GUCs may result in them being out of sync on segments", 'N'):
```
Comment (kalensk) [inline gpMgmt/bin/gpconfig:203]: If we haven't already, sync-up with @Nirali13 on final wording input/considerations.

File: gpMgmt/bin/gpconfig [hunk 3]
```diff
@@ -307,6 +308,13 @@ def do_change(options):
         LOGGER.error(msg)
         raise Exception(msg)
 
+    unreachable_hosts = []
+    if not options.coordinatoronly:
+        hosts = set(gp_array.get_hostlist(includeCoordinator=False))
+        unreachable_hosts = get_unreachable_segment_hosts(hosts, len(hosts))
+        if len(unreachable_hosts)>0:
```
Comment (kalensk) [inline gpMgmt/bin/gpconfig:315]: minor suggestion/nit: spaces around `>` for readability and consistent styling.

File: gpMgmt/test/behave/mgmt_utils/gpconfig.feature [hunk 4]
```diff
@@ -179,3 +179,39 @@ Feature: gpconfig integration tests
...
+    @demo_cluster
+    Scenario: gpconfig asks for confirmation when a host is unreachable and completes when user selects yes
+      Given the database is running
+        And the host for the primary on content 0 is made unreachable
+
+      When the user runs "gpconfig -c max_connections -v 100 -m 10 < test/behave/mgmt_utils/steps/data/yes.txt"
+
+      Then gpconfig should print "Host invalid_host is unreachable" to stdout
+        And gpconfig should print "Are you sure you want to ignore unreachable hosts?" to stdout
+        And gpconfig should return a return code of 0
```
Comment (kalensk) [inline gpMgmt/test/behave/mgmt_utils/gpconfig.feature:193]: Is there an "easy" way to confirm that this GUC actually got updated on the reachable hosts when the user proceeded? Similar idea for the other tests, but ensuring that the GUC did not get updated.

File: gpMgmt/test/behave/mgmt_utils/gpconfig.feature [hunk 5]
```diff
@@ -179,3 +179,39 @@ Feature: gpconfig integration tests
...
+      Given the database is running
+        And the host for the primary on content 0 is made unreachable
+
+      When the user runs "gpconfig -c max_connections -v 100 -m 10 < test/behave/mgmt_utils/steps/data/no.txt"
+
+      Then gpconfig should print "Host invalid_host is unreachable" to stdout
+        And gpconfig should print "Are you sure you want to ignore unreachable hosts?" to stdout
+        And gpconfig should print "User Aborted. Exiting." to stdout
+        And gpconfig should return a return code of 1
+
```
Comment (kalensk) [inline gpMgmt/test/behave/mgmt_utils/gpconfig.feature:207]: Since these tests are similar can we do a "table driven test". See other tests in this file as an example, but its basically one test with parameters passed in to run in different permutations. You use `Scenario Outline:` with an `Examples:` table with the variables which are referenced in the behave test as `<columnName>`.

File: gpMgmt/test/behave/mgmt_utils/gpconfig.feature (hunk 1 above)
Comment (divyeshddv) [reply gpMgmt/test/behave/mgmt_utils/gpconfig.feature]: Thanks, fixed

File: gpMgmt/bin/gpconfig (hunk 3 above)
Comment (divyeshddv) [reply gpMgmt/bin/gpconfig]: fixed

File: gpMgmt/test/behave/mgmt_utils/gpconfig.feature (hunk 4 above)
Comment (divyeshddv) [reply gpMgmt/test/behave/mgmt_utils/gpconfig.feature]: Changed behave tests to test this behavior. Now we are checking this through gpconfig -s, and from the postgresql.conf on each segment.

File: gpMgmt/test/behave/mgmt_utils/gpconfig.feature (hunk 5 above)
Comment (divyeshddv) [reply gpMgmt/test/behave/mgmt_utils/gpconfig.feature]: Refactored the tests for a table-driven tests

File: gpMgmt/bin/gpconfig (hunk 2 above)
Comment (niralis13) [reply gpMgmt/bin/gpconfig:203]: Suggestion: 

" One or more segment hosts are not currently reachable. If you continue with gpcofing, GUCs on unrechable segment host will not be updated. Do you want to continue? Yy|Nn (default=N):"

Comment (divyeshddv) [reply gpMgmt/bin/gpconfig]: Updated the warning/confirmation message

File: gpMgmt/bin/gpconfig [hunk 6]
```diff
@@ -307,25 +309,37 @@ def do_change(options):
...
+    if gp_array.standbyCoordinator is not None:
+        hosts.append(gp_array.standbyCoordinator.hostname)
+
+    if not options.coordinatoronly:
+        hosts.extend(gp_array.get_hostlist(includeCoordinator=False))
+
+    unreachable_hosts = get_unreachable_segment_hosts(hosts, len(hosts))
+    if len(unreachable_hosts) > 0:
+        confirm_user_wants_to_continue()
+
```
Comment (kalensk) [inline gpMgmt/bin/gpconfig:322]: If useful, consider a brief comment explaining that for --cordinatorOnly option we check coordinator and standby, and for all else we check all segments.

File: gpMgmt/bin/gpconfig [hunk 7]
```diff
@@ -307,25 +309,37 @@ def do_change(options):
...
 
             if not options.primaryvalue and not options.mirrorvalue:
-                do_add_config_script(pool, [seg for seg in gp_array.getSegDbList()], options.value, options)
+                do_add_config_script(pool, [seg for seg in gp_array.getSegDbList() if seg.hostname not in unreachable_hosts], options.value, options)
 
         # do the coordinator and standby
         if options.coordinatorvalue or options.remove:
-            do_add_config_script(pool, [seg for seg in [gp_array.coordinator, gp_array.standbyCoordinator] if seg is not None], options.coordinatorvalue, options)
+            do_add_config_script(pool, [seg for seg in [gp_array.coordinator, gp_array.standbyCoordinator] if seg is not None and seg.hostname not in unreachable_hosts],
+                                 options.coordinatorvalue, options)
```
Comment (kalensk) [inline gpMgmt/bin/gpconfig:342]: Rather than repeating `and seg.hostname not in unreachable_hosts` multiple times consider the following which not avoids the repetition, but seems easier to follow.

Create a local variable as follows
`reachable_hosts = set(gp_array.getSegDbList()).difference(unreachable_hosts)`
And update the various list comprehensions to use it such as:
`[seg for seg in reachable_hosts if seg.isSegmentPrimary()`

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 8]
```diff
@@ -1411,9 +1421,14 @@ def find_string_in_coordinator_data_directory(context, filename, output, escapeS
...
-        err_str = "Expected stdout string '%s' and found: '%s'" % (output, contents)
-        raise Exception(err_str)
+    if valuesShouldExist:
+        if not pat.search(contents):
+            err_str = "Expected stdout string '%s' and found: '%s'" % (output, contents)
+            raise Exception(err_str)
+    else:
+        if pat.search(contents):
+            err_str = "Did not expect stdout string '%s' but found: '%s'" % (output, contents)
+            raise Exception(err_str)
```
Comment (kalensk) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py:1431]: multiple nestings are hard to follow, consider the following instead:

```
if valuesShouldExist and not pat.search(contents):
   ...
if not valuesShouldExist and pat.search(contents):
   ...
```

Or remove the else and outdent the last if block.



Same concept for further down for 

```
if valuesShouldExist:
    if output not in actual:
        raise Exception('File %s on host %s does not contain "%s"' % (filepath, host, output))
 ```

**Review by kalensk — COMMENTED** [review]
Putting the "behave: refactor gpconfig behave tests steps" first makes things super clear to follow. This allows the main gpconfig commit to use it and makes things clear. Can you re-order them?

File: gpMgmt/bin/gpconfig (hunk 6 above)
Comment (divyeshddv) [reply gpMgmt/bin/gpconfig]: Added that as a comment, didn't find a better way to write this part of code.

File: gpMgmt/bin/gpconfig (hunk 7 above)
Comment (divyeshddv) [reply gpMgmt/bin/gpconfig]: Added this local variable as 
`reachable_segments = [seg for seg in gp_array.getSegDbList() and seg.hostname not in unreachable_hosts]`
since we need the segment info there, not just segment.hostname. Used it in comprehensions for easier readability.

Don't see an easy way to reduce the last use of 
`[seg for seg in [gp_array.coordinator, gp_array.standbyCoordinator] if seg is not None and seg.hostname not in unreachable_hosts]`

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 8 above)
Comment (divyeshddv) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: Updated to reduce nestings

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py [hunk 9]
```diff
@@ -1390,17 +1390,27 @@ def impl(context, filename, output):
...
 
-@then('verify that the last line of the file "{filename}" in the coordinator data directory contains the string "{output}" escaped')
-def impl(context, filename, output):
-    find_string_in_coordinator_data_directory(context, filename, output, True)
-
-
-@then('verify that the last line of the file "{filename}" in the coordinator data directory contains the string "{output}"')
-def impl(context, filename, output):
-    find_string_in_coordinator_data_directory(context, filename, output)
+@then('verify that the last line of the file "{filename}" in the coordinator data directory {contain} the string "{output}" escaped')
```
Comment (jmcatamney) [inline gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py:1393]: Nit: This step can be collapsed with the other one if you accept an optional parameter for the `escaped` at the end, e.g. change it to `...{contain} the string "{output}"{escape}` and pass `(escape == " escaped")` as the last parameter in `find_string_in_coordinator_data_directory`.

If you think that would be too cluttered, though, it's fine as-is.

**Review by jmcatamney — APPROVED** [review]

File: gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py (hunk 9 above)
Comment (divyeshddv) [reply gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py]: Updated it as suggested. I think it looks better now (instead of duplicating the same logic)
